### PR TITLE
refactor: middlewares configurable by specific app

### PR DESCRIPTION
**Pull request by stephanwlee**
Rationale: application.py did three things: (1) implement application
logic like /data/plugins_listing, (2) configured application with
plugins, and (3) configured middlewares.

Requirement: we would like to be able to apply different middlewares in
different order to configure how 404 gets handled and apply certain
application specific logic.

Changes:
Use `application.TensorBoardWSGI` as the base app that supplies feature
(1) and (2). `application.standard_tensorboard_wsgi` will have a default
middleware configurations appropriate for local TensorBoard.

Plans:
I plan on applying experiment_id middleware which is not used in local TensorBoard
in appropriate applications. I also plan on omitting path_prefix to the hosted version.

Tests:
- Ran application
- Ran application with path_prefix.

WANT_LGTM=all

### CODE REVIEW

File: tensorboard/backend/application.py [hunk 1]
```diff
@@ -161,11 +161,17 @@ def standard_tensorboard_wsgi(flags, plugin_loaders, assets_zip_provider):
       path_to_run = parse_event_files_spec(flags.logdir_spec)
     start_reloading_multiplexer(
         multiplexer, path_to_run, reload_interval, flags.reload_task)
-  return TensorBoardWSGIApp(
+
+  app = TensorBoardWSGIApp(
       flags, plugin_loaders, data_provider, assets_zip_provider, multiplexer)
+  app = empty_path_redirect.EmptyPathRedirectMiddleware(app)
```
Comment (wchargin) [inline tensorboard/backend/application.py]: Why is `experiment_id.ExperimentIdMiddleware` removed?

File: tensorboard/backend/application_test.py [hunk 2]
```diff
@@ -317,25 +318,29 @@ def testPluginEntryBadModulePath(self):
...
 
 
-class ApplicationBaseUrlTest(tb_test.TestCase):
+class StandardTensorBoardWSGIAppUrlTest(tb_test.TestCase):
   path_prefix = '/test'
   def setUp(self):
     plugins = [
-        FakePlugin(plugin_name='foo'),
-        FakePlugin(
+        FakePluginLoader(plugin_name='foo'),
```
Comment (wchargin) [inline tensorboard/backend/application_test.py]: I don’t follow; why are these changes required?

Comment (stephanwlee) [reply tensorboard/backend/application_test.py]: Answered offline but for posterity:

Previously, it was using TensorBoardWSGI which does not call `make_plugin_loader` as opposed to now, it uses `standard_tensorboard_wsgi` which calls `make_plugin_loader` through TensorBoardWSGIApp. `make_plugin_loader` requires plugins to be a class (not instance) of TBPlugin or an instance of TBPluginLoader.

**Comment by stephanwlee** [comment]
Thanks for suggesting the alternative.

```py
def _create_wsgi_app(self):
  ...
  app = empty_path_redirect.EmptyPathRedirectMiddleware(app)
  for middleware in self._experiment_id_middlewares:
    app = middleware(app)
  app = path_prefix.PathPrefixMiddleware(app, flags.path_prefix)
  ...
```
I did consider this option but I thought this was problematic for few reasons. Yes, we would like some global handler for errors and global validation of the security for all apps but (1) I want to be able to configure them programmatically and (2) I may want to swap and use other implementation of them. 
 Similarly different apps have different requirements that, for instance, I may want to omit `PathPrefixMiddleware` or add more in strategic places (before PathPrefix but after ExperimentId).

I am curious to learn @nfelt's opinion on this.
